journalbeat/reader/journal.go

```diff
@@ -0,0 +1,258 @@
+func New(c Config, done chan struct{}, state checkpoint.JournalState) (*Reader, error) {
+	f, err := os.Stat(c.Path)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to open file")
+	}
+
+	var j *sdjournal.Journal
+	if f.IsDir() {
+		j, err = sdjournal.NewJournalFromDir(c.Path)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to open journal directory")
+		}
+	} else {
+		j, err = sdjournal.NewJournalFromFiles(c.Path)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to open journal file")
+		}
+	}
+
+	r := &Reader{
+		j:       j,
+		changes: make(chan int),
+		config:  c,
+		done:    done,
+	}
+	r.seek(state.Cursor)
+
+	logp.Debug("reader", "New journal is opened for reading")
+
+	return r, nil
+}
+
+// NewLocal creates a reader to read form the local journal and moves the FP
+// to the configured position.
+func NewLocal(c Config, done chan struct{}, state checkpoint.JournalState) (*Reader, error) {
+	j, err := sdjournal.NewJournal()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to open local journal")
+	}
+
+	logp.Debug("reader", "New local journal is opened for reading")
+
+	r := &Reader{
+		j:       j,
+		changes: make(chan int),
+		config:  c,
+		done:    done,
+	}
+	r.seek(state.Cursor)
+	return r, nil
+}
+
+// seek seeks to the position determined by the coniguration and cursor state.
+func (r *Reader) seek(cursor string) {
+	if r.config.Seek == "cursor" {
+		if cursor == "" {
+			r.j.SeekHead()
+			logp.Debug("journal", "Seeking method set to cursor, but no state is saved for reader. Starting to read from the beginning")
+			return
+		}
+		r.j.SeekCursor(cursor)
+		_, err := r.j.Next()
+		if err != nil {
+			logp.Err("Error while seeking to cursor")
+		}
+		logp.Debug("journal", "Seeked to position defined in cursor")
+	} else if r.config.Seek == "tail" {
+		r.j.SeekTail()
+		logp.Debug("journal", "Tailing the journal file")
+	} else if r.config.Seek == "head" {
+		r.j.SeekHead()
+		logp.Debug("journal", "Reading from the beginning of the journal file")
+	}
+}
+
+// Follow reads entries from journals.
+func (r *Reader) Follow() <-chan *beat.Event {
+	out := make(chan *beat.Event)
+	go r.readEntriesFromJournal(out)
+
+	return out
+}
+
+func (r *Reader) readEntriesFromJournal(entries chan<- *beat.Event) {
+	defer close(entries)
+
+process:
+	for {
+		select {
+		case <-r.done:
+			return
+		default:
+			err := r.readUntilNotNull(entries)
+			if err != nil {
+				logp.Err("Unexpected error while reading from journal: %v", err)
+			}
+		}
+
+		for {
+			go r.stopOrWait()
+
+			select {
+			case <-r.done:
+				return
+			case e := <-r.changes:
+				switch e {
+				case sdjournal.SD_JOURNAL_NOP:
+					r.wait()
+				case sdjournal.SD_JOURNAL_APPEND, sdjournal.SD_JOURNAL_INVALIDATE:
+					continue process
+				default:
+					if e < 0 {
+						//logp.Err("Unexpected error: %v", syscall.Errno(-e))
+					}
+					r.wait()
+				}
+			}
+		}
+	}
+}
```

Comment:
I am not sure about the behavior of this method, see my comment concerning the range function.
It feels to me we should architecture this similar to how the harvester works, create a goroutine that watches a journald and send events to a channel and the channel is valid until we close the reader instead of creating a new channel per _follow call()_

Maybe there is something that I miss in the code or in the journald behavior.